trlx/models/modeling_nemo_sft.py

```diff
@@ -0,0 +1,523 @@
+        metrics, tables = zip(*outputs)
+        _, columns = tables[0]
+        rows = [r for trows, _ in tables for r in trows]
+
+        self.logger.log_text(key="samples", columns=columns, data=rows)
+
+        outputs_soa = {k: torch.as_tensor([d[k] for d in metrics]) for k in metrics[0].keys()}
+        # this assumes all validation microbatches are the same size
+        avg_outputs = {k: v.mean() for k, v in outputs_soa.items()}
+        for k, v in avg_outputs.items():
+            self.log(
+                f"val_metrics/{k}",
+                v,
+                prog_bar=True,
+                rank_zero_only=True,
+                sync_dist=True,
+            )
+
+    # Need to override this otherwise distributed fused adam won't work
+    # with frozen layers
+    def parameters(self):
+        return (p for p in self.model.parameters() if p.requires_grad)
+
+    def build_attention_mask_and_position_ids(
+        self, data: torch.LongTensor, reset_attention_mask: bool = False
+    ) -> Tuple[torch.BoolTensor, torch.LongTensor]:
+        micro_batch_size, seq_length = data.size()
+
+        # Attention mask (lower triangular).
+        if reset_attention_mask:
+            att_mask_batch = micro_batch_size
+        else:
+            att_mask_batch = 1
+        attention_mask = torch.tril(torch.ones((att_mask_batch, seq_length, seq_length), device=data.device)).view(
+            att_mask_batch, 1, seq_length, seq_length
+        )
+
+        # Position ids.
+        position_ids = torch.arange(seq_length, dtype=torch.long, device=data.device)
+        position_ids = position_ids.unsqueeze(0).repeat(micro_batch_size, 1)
+
+        # Convert attention mask to binary:
+        attention_mask = attention_mask < 0.5
+        return attention_mask, position_ids
+
+    def get_forward_output_and_loss_func(self, validation_step=False):
+        def fwd_output_and_loss_func(batch: List[torch.Tensor], model, checkpoint_activations_all_layers=None):
+            # On first and last pipeline stages, the input data is passed in
+            if parallel_state.get_pipeline_model_parallel_world_size() == 1:
+                input_ids, loss_mask = [b.cuda(non_blocking=True) for b in batch]
+                attention_mask, position_ids = self.build_attention_mask_and_position_ids(
+                    data=input_ids, reset_attention_mask=False
+                )
+            else:
+                input_ids, loss_mask, attention_mask, position_ids = None, None, None, None
+
+            output_tensor = model(
+                input_ids=input_ids,
+                position_ids=position_ids,
+                attention_mask=attention_mask,
+                checkpoint_activations_all_layers=checkpoint_activations_all_layers,
+            )
+
+            def loss_func(output_tensor):
+                # Shift logits and labels to align predictions
+                logits = output_tensor[:, :-1, :]
```

Comment:
When using SP, are the outputs already gathered somewhere within the model? Since otherwise the tokens at the end of the intermediate SP chunks will get lost (hence the gather_from_sequence_parallel_region in ILQL)